examples_tests/22.RaytracedAO/Renderer.cpp

```diff
@@ -616,6 +622,67 @@ void Renderer::initSceneNonAreaLights(Renderer::InitializationData& initData)
 		//initData.lightPDF.push_back(weight);
 		//initData.lights.push_back(light);
 	}
+
+	// if envmap texture is empty, create one with all values set to '_envmapBaseColor'
+	if(!m_envmap)
+	{
+		// Create CPUImageView
+		uint32_t width = 1u;
+		uint32_t height = 1u;
+
+		// Create CPUImage
+		ICPUImage::SCreationParams imgParams = {};
+		imgParams.type = asset::IImage::ET_2D;
+		imgParams.format = asset::EF_R32G32B32_SFLOAT;
+		imgParams.extent = VkExtent3D{width, height, 1u};
+		imgParams.mipLevels = 1u;
+		imgParams.arrayLayers = 1u;
+		imgParams.samples = asset::IImage::ESCF_1_BIT;
+		core::smart_refctd_ptr<ICPUImage> cpuEnvmapImage = ICPUImage::create(std::move(imgParams));
+				
+		auto texelBytesize = getTexelOrBlockBytesize(imgParams.format);
+		core::smart_refctd_ptr<ICPUBuffer> buffer = core::make_smart_refctd_ptr<ICPUBuffer>(texelBytesize * width * height);
+		
+		auto regionsDynArray = core::make_refctd_dynamic_array<core::smart_refctd_dynamic_array<IImage::SBufferCopy>>(1u);
+		auto regions = regionsDynArray->begin();
+		regions[0].bufferOffset = 0u;
+		regions[0].bufferRowLength = imgParams.extent.width; // could round up to multiple of 8 bytes in the future
+		regions[0].bufferImageHeight = imgParams.extent.height;
+		regions[0].imageSubresource.aspectMask = static_cast<IImage::E_ASPECT_FLAGS>(0u);
+		regions[0].imageSubresource.mipLevel = 0u;
+		regions[0].imageSubresource.baseArrayLayer = 0u;
+		regions[0].imageSubresource.layerCount = 1u;
+		regions[0].imageOffset = { 0u,0u,0u };
+		regions[0].imageExtent = imgParams.extent;
+
+		cpuEnvmapImage->setBufferAndRegions(std::move(buffer), std::move(regionsDynArray));
+
+		// Fill CPUImage with color
+		CFillImageFilter::state_type fill;
+		fill.subresource.aspectMask = asset::IImage::EAF_COLOR_BIT;
+		fill.subresource.mipLevel = 0u;
+		fill.subresource.baseArrayLayer = 0u;
+		fill.subresource.layerCount = 1u;
+		fill.outRange = { {0u,0u,0u}, {width, height, 1u} };
+		fill.outImage = cpuEnvmapImage.get();
+		fill.fillValue.asFloat = _envmapBaseColor;
+		if (!CFillImageFilter::execute(&fill))
+			std::cout << "CFillImageFilter failed for envmap color filling." << std::endl;
+		
+		ICPUImageView::SCreationParams viewParams;
+		viewParams.flags = static_cast<ICPUImageView::E_CREATE_FLAGS>(0u);
+		viewParams.image = cpuEnvmapImage;
+		viewParams.format = viewParams.image->getCreationParameters().format;
+		viewParams.viewType = IImageView<ICPUImage>::ET_2D;
+		viewParams.subresourceRange.baseArrayLayer = 0u;
+		viewParams.subresourceRange.layerCount = 1u;
+		viewParams.subresourceRange.baseMipLevel = 0u;
+		viewParams.subresourceRange.levelCount = 1u;
+
+		auto cpuEnvmapImageView = ICPUImageView::create(std::move(viewParams));
+		m_envmap = m_driver->getGPUObjectsFromAssets(&cpuEnvmapImageView.get(), &cpuEnvmapImageView.get() + 1u)->front();
+	}
```

Comment:
I want to step up the game, and get rid of `m_staticViewData.envmapBaseColor`

just add it to every pixel in the envmap... profit, you only shade with envmap ;)